src/assisted_installer_controller/assisted_installer_controller.go

```diff
@@ -177,6 +177,31 @@ func logHostsStatus(log logrus.FieldLogger, hosts map[string]inventory_client.Ho
 	log.Infof("Hosts status: %v", hostsStatus)
 }
 
+func (c *controller) updateNodesLabels() bool {
+	ignoreStatuses := []string{models.HostStatusDisabled}
+	ctxReq := utils.GenerateRequestContext()
+	log := utils.RequestIDLogger(ctxReq, c.log)
+
+	assistedNodesMap, err := c.ic.GetHosts(ctxReq, log, ignoreStatuses)
+	if err != nil {
+		log.WithError(err).Error("Failed to get node map from the assisted service")
+		return KeepWaiting
+	}
+
+	for hostname, hostData := range assistedNodesMap {
+		if len(hostData.Host.NodeLabels) == 0 {
+			continue
+		}
+		for ok := true; ok; ok = err != nil {
+			err = c.kc.PatchNodeLabels(hostname, hostData.Host.NodeLabels)
+			if err != nil {
+				log.WithError(err).Errorf("Failed to patch node %s with node labels %s", hostname, hostData.Host.NodeLabels)
```

Comment:
You'll still falling through to `ExitWaiting`, there's no retrying going on here

Reply:
This is a do-while loop, so, when this statement `err!=nil` is true i.e we encounter an error, the patch function would be executed again.

Reply:
Completely missed that `for` loop. Please use the KeepWaiting mechanism to try again if any of the hosts fail a label update, and only do ExitWaiting if all of them succeeded, this retry-loop inside an already retrying function is redundant.

Also it's a bit weird to fixate on one host, I think it's more neat to go through all of them, skip the ones that fail, and then retry all of the ones that failed in the next run. This way if there's just one bad host it would be more obvious that it's the only bad one, because we can see that rest got their labels correctly.

Also `for ok := true; ok; ok = err != nil` gave me a headache, I don't even want to wrap my head around it. Shouldn't it be `for needRetry := true; needRetry; needRetry = err != nil`? ok is a bit misleading, because it sounds like it means everything is fine while in practice is should mean "please keep looping, there's an error"